### PR TITLE
Migrate Theme Catalog Page of Design Section

### PR DESCRIPTION
**Pull request by michaelKaefer**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migrate AdminThemesCatalogueController to ThemeController
| Type?         | improvement
| Category?     | CO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | -
| How to test?  | Admin > Improve > Design > Themes Catalog

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9154)

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
Hello @michaelKaefer,

sounds good to me.

About your git history, let me fix it for you. Time for review :)

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace PrestaShopBundle\Controller\Admin\Improve\Design;
+
+use PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Displays themes from Addons under "Improve > Design > Themes Catalog"
+ *
+ * @author Michael Käfer <michael.kaefer1@gmx.at>
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: we usually don't add author name in classes ... but as you're the very first one community contributor, I'm in favor of keeping it.

WDYT @eternoendless ?

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ */
+class ThemeController extends FrameworkBundleAdminController
+{
+    /**
+     * @var string The legacy controller name for routing.
+     */
+    const CONTROLLER_NAME = 'AdminThemesCatalog';
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: This constant is not required anymore, sorry for that docs are outdated already.

**Comment by michaelKaefer** [comment]
Hi @mickaelandrieu 
I do not understand - but thank you anyway :smile:

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    /**
+     * Displays themes from Addons under "Improve > Design > Themes Catalog"
+     *
+     * @param Request $request
+     *
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function indexAction(Request $request)
+    {
+        $pageContent = file_get_contents('https://addons.prestashop.com/iframe/search-1.7.php?'.
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: How about:

```php
$context = $this->getContext();
$configuration = $this->get('prestashop.adapter.legacy.configuration');

$iframeURI = 'https://addons.prestashop.com/iframe/search-1.7.php' 
    . http_build_query([
        'psVersion' => $configuration->get('_PS_VERSION_'),
        'isoLang' => $context->language->iso_code,
        'isoCurrency' => $context->currency->iso_code,
        'isoCountry' => $this->getContext()->country->iso_code,
        'activity' => $configuration->getInt('PS_SHOP_ACTIVITY'),
        'parentUrl' => $request->getSchemeAndHttpHost(),
        'onlyThemes' => 1
    ])
```

* with `http_build_query` you don't need to care about "&" and "?" anymore,
* we have some helpers in 1.7 with Configuration like `getInt()` and `getBoolean()`

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            'requireAddonsSearch' => true,
+            'requireBulkActions' => false,
+            'showContentHeader' => true,
+            'enableSidebar' => true,
+            'help_link' => $this->generateSidebarLink('AdminThemesCatalog'),
+            'requireFilterStatus' => false,
+            'level' => $this->authorizationLevel($this::CONTROLLER_NAME),
+        ]);
+    }
+}
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: please add an extra line at the end of the file.

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            'pageContent' => $pageContent,
+            'layoutHeaderToolbarBtn' => [],
+            'layoutTitle' => $this->trans('Themes Catalog', 'Admin.Navigation.Menu'),
+            'requireAddonsSearch' => true,
+            'requireBulkActions' => false,
+            'showContentHeader' => true,
+            'enableSidebar' => true,
+            'help_link' => $this->generateSidebarLink('AdminThemesCatalog'),
+            'requireFilterStatus' => false,
+            'level' => $this->authorizationLevel($this::CONTROLLER_NAME),
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: you can remove `level` property, you don't need it (and it should be removed from modules list iframe too, in a future contribution)

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        $pageContent = file_get_contents('https://addons.prestashop.com/iframe/search-1.7.php?'.
+            'psVersion='.$this->get('prestashop.adapter.legacy.configuration')->get('_PS_VERSION_').'&'.
+            'isoLang='.$this->getContext()->language->iso_code.'&'.
+            'isoCurrency='.$this->getContext()->currency->iso_code.'&'.
+            'isoCountry='.$this->getContext()->country->iso_code.'&'.
+            'activity='.(int) $this->get('prestashop.adapter.legacy.configuration')->get('PS_SHOP_ACTIVITY').'&'.
+            'parentUrl='.$request->getSchemeAndHttpHost().'&'.
+            'onlyThemes=1');
+
+        return $this->render('@PrestaShop/Admin/Improve/Design/Theme/addons_store.html.twig', [
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]:
```diff
- @PrestaShop/Admin/Improve/Design/Theme/addons_store.html.twig
+ @PrestaShop/Admin/Improve/Design/ThemesCatalogPage/addons_store.html.twig
```

and move the template to the right location.

File: src/PrestaShopBundle/Resources/config/routing/admin/routing_theme.yml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+admin_theme_catalog:
+    path: /addons-store
+    methods: [GET]
+    defaults:
+        _controller: 'PrestaShopBundle:Admin\Improve\Design\Theme:index'
+        _legacy_controller: AdminThemesCatalog
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/config/routing/admin/routing_theme.yml]: also add an extra empty line at the end of the file

File: src/PrestaShopBundle/Resources/config/routing/routing_admin.yml [hunk 8]
```diff
@@ -35,6 +35,10 @@ _admin_supplier_routing:
     resource: "admin/routing_supplier.yml"
     prefix: /supplier
 
+_admin_theme_routing:
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/config/routing/routing_admin.yml:38]: :+1:

**Review by mickaelandrieu — CHANGES_REQUESTED** [review]
some minor changes but it's already working good!

Thanks @michaelKaefer

**Comment by michaelKaefer** [comment]
Thank you @mickaelandrieu again! Made another commit :)

**Comment by mickaelandrieu** [comment]
> I do not understand - but thank you anyway smile

We shouldn't have these commits:

git_history

as they are not related to this contribution.

I'll clean up before merge :)

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by mickaelandrieu** [comment]
Dear @marionf, I want you to check this great contribution as soon as possible! :tada:

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by mickaelandrieu** [comment]
@michaelKaefer tbh I don't know for the `@author` notation, I don't see all of them in PrestaShop code... 

This is something we need to discuss with @eternoendless and document :)

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+     * @param Request $request
+     *
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function indexAction(Request $request)
+    {
+        $context = $this->getContext();
+        $configuration = $this->get('prestashop.adapter.legacy.configuration');
+
+        $pageContent = file_get_contents('https://addons.prestashop.com/iframe/search-1.7.php'
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: Can `1.7` be a parameter or something related to the current PS_VERSION? :)

**Comment by marionf** [comment]
Hello @michaelKaefer 

I have a page not found when I click on Design > Themes Catalog because it's still the old link

capture du 2018-06-01 09-39-48

**Comment by mickaelandrieu** [comment]
Hi! To update the menu link, what you need to do is to update the class `Link` https://github.com/PrestaShop/prestafony-project/blob/master/docs/migration-guide.md#routing-in-prestashop, this will use the Symfony router for the menu item.

Nice spotted @marionf  :)

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php (hunk 9 above)
Comment (michaelKaefer) [reply src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: @PierreRambaud I don't know, just copied the URL from the old controller. I also would suppose it's the version and made another commit :+1:

**Comment by michaelKaefer** [comment]
Ups, stupid! Thank you @marionf and @mickaelandrieu ! I fixed it :+1:

**Review by mickaelandrieu — APPROVED** [review]

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace PrestaShopBundle\Controller\Admin\Improve\Design;
+
+use PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Displays themes from Addons under "Improve > Design > Themes Catalog"
+ */
+class ThemeController extends FrameworkBundleAdminController
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php]: one more update: I prefer ``ThemeCatalogController``, you also have to update the routing, and then we're good

**Review by mickaelandrieu — CHANGES_REQUESTED** [review]

File: src/PrestaShopBundle/Resources/config/routing/admin/routing_theme.yml [hunk 11]
```diff
@@ -0,0 +1,6 @@
+admin_theme_catalog:
+    path: /addons-store
+    methods: [GET]
+    defaults:
+        _controller: 'PrestaShopBundle:Admin\Improve\Design\Theme:index'
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/config/routing/admin/routing_theme.yml]:
```diff
- 'PrestaShopBundle:Admin\Improve\Design\Theme:index'
+ 'PrestaShopBundle:Admin\Improve\Design\ThemeCatalog:index'
```

Comment (michaelKaefer) [reply src/PrestaShopBundle/Resources/config/routing/admin/routing_theme.yml]: @mickaelandrieu ok, thanks! I made another commit :+1:

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by michaelKaefer** [comment]
@mickaelandrieu Sorry Mickael, I forgot to rename the file. Done now.

**Review by Matt75 — COMMENTED** [review]
Hi everyone
I’m not sure inject external content directly to the DOM is a good things... What’s happening if Prestashop server where compromised ?
Iframe did not provide some interesting security about externals contents ? Like sandboxing etc...
I’m not an expert but currently the new way choosed here seems to me risky.

Why don’t use VueJs to create a view to display content extracted from a json/text return from Addons call ?

**Comment by mickaelandrieu** [comment]
@Matt75 if PrestaShop server is compromised, using Vuejs from json content sent by PrestaShop server won't help here.

Theme Calogue and ModuleCatalogue pages are managed by a different team of PrestaShop (Addons), and the content is really dynamic (special offers, events, etc...) this is why it's not managed directly in GitHub and not managed in our code. For now you can see it as a simple list of themes displayed but they can make this page evolve the way they want to fit some marketing needs for instance.

**Comment by mickaelandrieu** [comment]
@Matt75 I don't know ... this is something you can probably ask to PrestaShop Twitter account.

Note that here, we're only migrating 1.6 old content to Symfony framework without altering the behavior: so we're not introducing new trackers or others things, we only change the way we get the user request and returns to him/her a response.

**Comment by Matt75** [comment]
Right, moreover we can modify the code before use it in production to not execute external call to Prestashop but this require some skills.
I just want to say I’m not good with that because I think this is not really RGPD compliant because of lack information.
I don’t like that

**Comment by mickaelandrieu** [comment]
@Matt75 duly noted, but will be addressed in another contribution: we migrate to Symfony with exactly the same features. If the feature doesn't comply the RGPD - that I doubt to be honest - this need to be reviewed by the Product team.

The best you can do here is to submit an issue to the forge :)

Thanks for your understanding :)

**Comment by PierreRambaud** [comment]
I'm agree with @mickaelandrieu, I'm sure it's GDPR compliant because it's just a http request to retrieve modules, not to store any personal data. It's the same when you use a cdn to display bootstrap css =)

**Comment by Matt75** [comment]
My problem is if Prestashop server is compromised or if the dns is altered or domain expired or if Prestashop Addons add some JavaScript or tracker to the code merged with return content from add on to the DOM directly it can be an injection potentially problematic.
Here I have no guaranty about this content directly injected to the dom is safe, validated etc...
I already have damage with a same issue in an other project that why I am not secured with this behavior
I prefere external content being filtered, sanitized and validated before.
I have no problem with the fact to retrieve content from Addons or the fact of technical data is sended for stats & analysis, i’m just not in favor of this behavior

**Comment by marionf** [comment]
Hello @michaelKaefer 

There are some design issues:

capture du 2018-06-04 09-32-29

capture du 2018-06-04 09-32-45

capture du 2018-06-04 09-33-38

Before:

capture du 2018-06-04 09-38-43

capture du 2018-06-04 09-35-05

capture du 2018-06-04 09-37-17

**Comment by mickaelandrieu** [comment]
@marionf are these issues already available in `develop` branch?

Design comes from Addons as far as I know :/

**Comment by PierreRambaud** [comment]
If design comes from Addons and we inject it directly in the DOM without frames, we need to fix our css to be compatible with :)

**Comment by mickaelandrieu** [comment]
@marionf I don't reproduce on his branch, any chance this is about web browser cache or something related?

modules_catalog

**Comment by marionf** [comment]
@mickaelandrieu I'am in private navigation

On develop branch  it's ok :

capture du 2018-06-04 16-46-27

On the PR :

capture du 2018-06-04 16-50-10

**Comment by mickaelandrieu** [comment]
@marionf ok I can reproduce it, I was on Module catalog page and not on Theme Catalog page :+1:

**Comment by michaelKaefer** [comment]
Thank you @marionf ! I will try on friday.

**Comment by mickaelandrieu** [comment]
@michaelKaefer you can't change the design of the page as the content comes from an iframe.

For me, your work is done here, and we need to update the iframe from our side.

What do you think @marionf ?

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by marionf** [comment]
It's ok for me :)

File: classes/Link.php [hunk 12]
```diff
@@ -743,21 +743,14 @@ public function getAdminLink($controller, $withToken = true, $sfRouteParams = ar
                     'AdminModulesManage' => 'admin_module_manage',
                     'AdminModulesNotifications' => 'admin_module_notification',
                     'AdminModulesSf' => 'admin_module_manage',
+                    'AdminCustomerPreferences' => 'admin_customer_preferences',
```
Comment (PierreRambaud) [inline classes/Link.php]: Duplicate line 737

**Review by PierreRambaud — DISMISSED** [review]
Duplicate line

File: classes/Link.php (hunk 12 above)
Comment (mickaelandrieu) [reply classes/Link.php]: not anymore :)

Comment (PierreRambaud) [reply classes/Link.php]: Lines were sorted and you broke it :crying_cat_face:

**Review by mickaelandrieu — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
Thanks @michaelKaefer, do you want to share your feelings about this first migration?

We'd like to onboard more people like you, so your feedback is really important to us.

Also, this is only the start of the Open Source journey for you if you want to discover PrestaShop 1.7, let me know and I can present you some tasks I'm working on right now for 1.7.5.

Also, if you have ideas of improvements, they are very welcome!

Thanks for your time and your commitment, you're awesome :-)

**Comment by kpodemski** [comment]
Well done @michaelKaefer 

and i'm glad to see that team is willing to accept this kind of PR here, me and many other developers were completely ignored many times in the past after submitting more complex PRs

**Comment by mickaelandrieu** [comment]
@kpodemski time for you to contribute again maybe? If you want to try to migrate a page to learn how PrestaShop 1.7 back office work under the hood, I'll help you 👍

**Comment by kpodemski** [comment]
@mickaelandrieu yes, i'm back at it :) this is why i'm digging through all recent PRs to see what is going on currently :)

**Comment by michaelKaefer** [comment]
@kpodemski I'm new to contributing to PS. But in this short period quite some people where really helpful, especially @mickaelandrieu :+1: 

@mickaelandrieu My feelings: an easy controller to migrate, you have choosen a perfect start for somebody new. I really like the decision to rely on Symfony concerning all the not-webshop stuff and it's cool to contribute to open source - even if it's a small part. So, thank you for your effort and patience, I really appreciate it :)

@mickaelandrieu Yeah, I'm ready for tasks, do you plan to maintain a list or something where we can choose?

**Comment by mickaelandrieu** [comment]
There is one roadmap here => https://github.com/PrestaShop/prestafony-project/issues/10 this is what we'd like to do concerning the migration of back office (our top priority, for both us and the community).

What we can do now is working on a harder controller together, what do you think? Take a look at the list and let me know if something sounds interesting for you ;) good night !

**Pull request merged**